libcommon/dpcommon/threading_win32.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2022
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "common.h"
+#include "threading.h"
+#include <windows.h>
+
+static void free_error_state();
+
+DP_Mutex *DP_mutex_new(void)
+{
+    return (DP_Mutex *)CreateMutexA(NULL, FALSE, NULL);
+}
+
+void DP_mutex_free(DP_Mutex *mutex)
+{
+    CloseHandle((HANDLE *)mutex);
+}
+
+bool DP_mutex_lock(DP_Mutex *mutex)
+{
+    DP_ASSERT(mutex);
+    WaitForSingleObject((HANDLE *)mutex, INFINITE);
+    return true;
+}
+
+DP_MutexResult DP_mutex_try_lock(DP_Mutex *mutex)
+{
+    DP_ASSERT(mutex);
+    if (WaitForSingleObject((HANDLE *)mutex, 0) == WAIT_OBJECT_0) {
+        return DP_MUTEX_OK;
+    }
+    else {
+        return DP_MUTEX_BLOCKED;
+    }
+}
+
+bool DP_mutex_unlock(DP_Mutex *mutex)
+{
+    DP_ASSERT(mutex);
+    ReleaseMutex((HANDLE *)mutex);
+    return true;
+}
+
+
+DP_Semaphore *DP_semaphore_new(unsigned int initial_value)
+{
+    return (DP_Semaphore *)CreateSemaphoreW(NULL, initial_value, LONG_MAX,
+                                            NULL);
+}
+
+void DP_semaphore_free(DP_Semaphore *sem)
+{
+    DP_ASSERT(sem);
+    CloseHandle(sem);
+}
+
+int DP_semaphore_value(DP_Semaphore *sem)
+{
+    DP_ASSERT(sem);
+
```

Comment:
Maybe comment the reason for this: acquire and release is necessary because you can't get the value of a Win32 semaphore directly.

Also, incredibly minor point: this is the only if that uses a Yoda condition, all the others do it the normal way round.